### PR TITLE
feat: convert projects to Playwright

### PR DESCRIPTION
**Pull request by Sembauke**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #XXXXX

### CODE REVIEW

**Comment by Sembauke** [comment]
> The tests are still failing because there is a stray } in the editor, causing the code to be invalid and blocking the submission. The solution code in the JSON file looks correct to me, so I'm still (pulling my hair) trying to figure out where that character comes from (both editor.evaluate() and editor.fill() gave the same result).

You deleted the parameter which disables bracket and quote completion in the editor.

image

This function has been added as even when filling or pasting code the editor completes brackets:

```
    // disable bracket completion when testing as it duplicates brackets when filling text
    const disableBracketCompletion =
      window.location.href.includes('testing=true');

    console.log(window.location.href, disableBracketCompletion);

    if (disableBracketCompletion) {
      editor?.updateOptions({
        autoClosingBrackets: 'never',
        autoClosingQuotes: 'never'
      });
    }
```

I should have left a comment about it.

**Comment by ojeytonwilliams** [comment]
@huyenltnguyen yep, it's the cookies. The issue is that the cookies for `certifieduser` are used even after the db is seeded with `developmentuser`. The fix/workaround is to create storageStates for both so that you can switch between users.

I'll create a quick PR for that, so you can use it here.

**Comment by ojeytonwilliams** [comment]
https://github.com/freeCodeCamp/freeCodeCamp/pull/54752 should help.

Once it's in, switching users means both changing storage state and seeding the desired user.

**Comment by ojeytonwilliams** [comment]
> I think the second test.use runs before the tests and overrides the first test.use, which is why the conflict happened

Pretty much. It's not exactly a conflict, it's just that multiple `test.use` calls override each other. As far as I understand it, `test.use` simply set the Playwright config for that scope (the entire file or a single describe block), so making two calls in the same scope is equivalent to just making the second call.

File: e2e/projects.spec.ts [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+    }
+  });
+});
+
+test.describe('JavaScript projects can be submitted and then viewed in /settings and on the certifications', () => {
+  test.use({ storageState: 'playwright/.auth/development-user.json' });
+  test.beforeAll(() => {
+    execSync('node ./tools/scripts/seed/seed-demo-user');
+  });
+
```
Comment (ojeytonwilliams) [inline e2e/projects.spec.ts:91]:
```suggestion
  test.skip(
    ({ browserName }) => browserName !== 'chromium',
    'Only chromium allows us to use the clipboard API.'
  );

```
So we can paste below.

File: e2e/projects.spec.ts [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+    expectedPaths.push('/learn/javascript-algorithms-and-data-structures/');
+
+    const contents = projectsInOrder[0].solutions[0][0].contents as string;
+
+    // Test the direct flow once.
+    // Use the `testing=true` parameter to prevent the editor
+    // from automatically adding a closing bracket.
+    await page.goto(
+      '/learn/javascript-algorithms-and-data-structures/javascript-algorithms-and-data-structures-projects/palindrome-checker?testing=true'
```
Comment (ojeytonwilliams) [inline e2e/projects.spec.ts:140]:
```suggestion
      '/learn/javascript-algorithms-and-data-structures/javascript-algorithms-and-data-structures-projects/palindrome-checker'
```
It won't be necessary if we use the clipboard.

File: e2e/projects.spec.ts [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+    const editor = await getProjectEditors({ page, isMobile });
+    await focusProjectEditor({ page, browserName, isMobile });
+    await clearProjectEditor({ page, browserName });
+
+    // To pass on firefox, we need to make sure the editor value is only set once
+    // instead of multiple times. (Bug in firefox)
+    await editor.evaluate((element, value) => {
+      (element as HTMLTextAreaElement).value = value;
+      element.dispatchEvent(new Event('input', { bubbles: true }));
+    }, contents);
```
Comment (ojeytonwilliams) [inline e2e/projects.spec.ts:152]:
```suggestion
    await page.evaluate(
      async contents => await navigator.clipboard.writeText(contents),
      contents
    );
    // Only runs on Chromium, so Control will work.
    await page.keyboard.press('Control+v');
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

**Review by huyenltnguyen — APPROVED** [review]
Hookay, I think the PR is good to go.

This is one heck of a test suite to write / rewrite 😰

File: e2e/utils/request.ts [hunk 4]
```diff
@@ -3,24 +3,35 @@ import { APIRequestContext, expect } from '@playwright/test';
...
     {
       data,
       headers: { 'csrf-token': csrfToken! }
     }
   );
+
+  if (res.status() !== 200) {
+    console.log(await res.json());
+  }
+
```
Comment (ojeytonwilliams) [inline e2e/utils/request.ts:34]:
```suggestion
```
We can just look in the network tab if there's an error.

File: client/src/templates/Challenges/classic/editor.tsx [hunk 5]
```diff
@@ -309,6 +309,7 @@ const Editor = (props: EditorProps): JSX.Element => {
       props.challengeType !== challengeTypes.multifilePythonCertProject
         ? 2
         : 4,
+
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/classic/editor.tsx:312]:
```suggestion
```

**Review by ojeytonwilliams — APPROVED** [review]
This was a struggle. Nice work, @Sembauke and @huyenltnguyen!

LGTM :+1:

**Pull request merged**